mongodbatlas/fw_common.go

```diff
@@ -0,0 +1,68 @@
+package mongodbatlas
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+// common struct for all framework resources. Implements the following plugin-framework defined functions:
+// - Metadata
+// - Configure
+// client is left empty and populated by the framework when envoking Configure method.
+// resourceName must be defined when creating an instance of a resource.
+type RSCommon struct {
+	client       *MongoDBClient
+	resourceName string
+}
+
+func (r *RSCommon) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, r.resourceName)
+}
+
+func (r *RSCommon) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	client, err := ConfigureClient(req.ProviderData)
+	if err != nil {
+		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
+		return
+	}
+	r.client = client
+}
+
+// common struct for all framework data sources. Implements the following plugin-framework defined functions:
+// - Metadata
+// - Configure
+// client is left empty and populated by the framework when envoking Configure method.
+// dataSourceName must be defined when creating an instance of a data source.
+type DSCommon struct {
+	client         *MongoDBClient
+	dataSourceName string
+}
+
+func (d *DSCommon) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, d.dataSourceName)
+}
+
+func (d *DSCommon) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	client, err := ConfigureClient(req.ProviderData)
+	if err != nil {
+		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
+		return
+	}
+	d.client = client
+}
+
+func ConfigureClient(providerData any) (*MongoDBClient, error) {
```

Comment:
any chance we could avoid using `any`?

Reply:
Not in this case. The ConfigureRequest struct defined by terraform plugin framework defines the `ProviderData` field as `any` becuase the data is provider-specifc.